### DISCUSSION TITLE
Running ServiceWorker in Capacitor App on iOS

### DISCUSSION THREAD

**Discussion by merri-ment** [ionic-team/capacitor#5502] [Q&A]
### tl;dr

Essentially i need to set COOP/COEP headers to enable SharedArrayBuffer functionality in my Capacitor v3 iOS App and running a ServiceWorker to set the headers is the only solution i have found so far.

### Background to the problem
Both Chrome Android and Safari iOS [now support SharedArrayBuffer](https://caniuse.com/?search=SharedArrayBuffer) which now means Ffmpeg.wasm can utilise multi-threading. SharedArrayBuffer requires a secure context to enable functionality, this is done by setting the following headers;

> Cross-Origin-Embedder-Policy: require-corp
> Cross-Origin-Opener-Policy: same-origin

Further requirements;

- The site needs to be served over HTTPS or localhost
- And the headers have to be set server side (headers cannot be set using meta http-equiv=".." in the head tag)

I've found this problematic in Capacitor to implement. 
My App is self contained (App files served locally), it doesn't require external http requests to serve the App files. This means setting the required headers server-side isn't an option.

### The Work-around solution
The following solution suggests using a ServiceWorker to set the COOP/COEP headers client-side, which is exactly what i need. [Enabling COOP/COEP without touching the server](https://dev.to/stefnotch/enabling-coop-coep-without-touching-the-server-2d3n)

### The Problem
ServiceWorkers have their own set of requirements and this is where i'm stuck.

To enable this ServiceWorker API on iOS i'm using [App-Bound Domains](https://webkit.org/blog/10882/app-bound-domains/).

The following issue in github [Support deploying a PWA with capacitor as native app](https://github.com/quasarframework/quasar/issues/10934) outlines the steps required to implement App-Bound Domains in Capacitor.

Which include;
- Adding a list of your domains incl. server.hostname / localhost to Info.plist.
- Enabling limitsNavigationsToAppBoundDomains in your src - capacitor/capacitor.config.json

After setting App-Bound Domains i get the following error;
> serviceWorker.register() must be called with a script URL whose protocol is either HTTP or HTTPS

This happens because Capacitor apps are served from capacitor:// in iOS.
Is there anyway around this? ( that does not require serving from a remote url).

Or has anyone discovered a better approach to setting COOP/COEP headers client-side in Capacitor?

Thanks Cam

**Comment by frissonlabs** [comment]
Hi @merri-ment , did you end up solving this? Running into the same obstacle as well!

**Reply by frissonlabs** [reply]
For anyone else, here's a barebones test app for demoing this issue:
[https://github.com/frissonlabs/ionic-react-capacitor-sharedarraybuffer](https://github.com/frissonlabs/ionic-react-capacitor-sharedarraybuffer)

SharedArrayBuffer is undefined in Android even though the COOP/COEP headers are set, not sure why this doesn't work.

**Comment by pwespi** [comment]
COOP/COEP heaaders are required for crossOriginIsolated capabilities, but they are not necessarily sufficient.
See https://github.com/whatwg/html/pull/6098

**Comment by galaxyblur** [comment]
Also facing this issue and would appreciate any help from the Capacitor team or community! Particularly, we should be able to tap into the Capacitor http server and set headers as needed.

**Comment by marcelo-nagatomo** [comment]
hey all, I'm also facing this issue while trying to enable `SharedArrayBuffer` on `capacitor-ios`.. 
I kind of duplicated this issue here: https://github.com/ionic-team/capacitor/issues/6182
And here are some findings I got while debugging in iOS: https://github.com/ionic-team/capacitor/issues/6182#issuecomment-1362224897

I tested in Android Chrome opening the link below and `SharedArrayBuffer` seems to be available:
https://cross-origin-isolation.glitch.me/?coep=require-corp&coop=same-origin

<img width="1209" alt="Screen Shot 2022-12-22 at 5 03 40 pm" src="https://user-images.githubusercontent.com/64799803/209217119-136ae4d8-30c6-46e5-a719-b556714ee4ef.png">

**Reply by marcelo-nagatomo** [reply]
It seems `Android Chrome` and `Android Webview` are 2 different things.... =/
`Android WebView` doesn't support site isolation so there's no way to enable `SharedArrayBuffer` on the webview today
More info:
https://chromium.googlesource.com/chromium/src/+/main/docs/process_model_and_site_isolation.md#modes-and-availability

On iOS I was able to downgrade to `cordova-plugin-ionic-webview@2.5.3` which uses GCDWebServer rather than WebKit's URLSchemeHandler. After changing its source code to add COEP/COOP headers I was able to enable cross-origin isolation.

A proper solution would be a new plugin that uses a local web server or wait to see if WebKit could one day allow the URLSchemeHandler to set response headers on local files.

**Comment by skijit** [comment]
@marcelo-nagatomo Can you clarify how you were able to downgrade to a different webview in a capacitor project?  My initial tests have failed and it is listed as a [known incompatible plugin](https://capacitorjs.com/docs/v2/cordova/known-incompatible-plugins) in the capacitor documentation.

**Reply by patrickcorrigan** [reply]
Did you make any progress on this @skijit :)
